whisk/action_test.go

```diff
@@ -0,0 +1,151 @@
+)
+
+const (
+    NODE_ACTION_NO_CODE = `{
+        "name": "test",
+        "publish": false,
+        "annotations": [
+            {
+                "key": "exec",
+                "value": "nodejs:6"
+            }
+        ],
+        "version": "0.0.1",
+        "exec": {
+            "kind": "nodejs:6",
+            "binary": false
+        },
+        "parameters": [],
+        "limits": {
+            "timeout": 60000,
+            "memory": 256,
+            "logs": 10
+        },
+        "namespace": "test@openwhisk"
+    }`
+
+    NODE_ACTION = `{
+        "name": "test",
+        "publish": false,
+        "annotations": [
+            {
+                "key": "exec",
+                "value": "nodejs:6"
+            }
+        ],
+        "version": "0.0.1",
+        "exec": {
+            "kind": "nodejs:6",
+            "code": "...",
+            "binary": false
+        },
+        "parameters": [],
+        "limits": {
+            "timeout": 60000,
+            "memory": 256,
+            "logs": 10
+        },
+        "namespace": "test@openwhisk"
+    }`
+)
+
+type ActionResponse struct {
+    Body    string
+}
+
+type ActionRequest struct {
+    Method  string
+    URL     string
+}
+
+var actionResponse = &ActionResponse{}
+var actionRequest = &ActionRequest{}
+
+type MockClient struct {}
+
+func (c *MockClient) NewRequestUrl(method string, urlRelResource *url.URL, body interface{}, includeNamespaceInUrl bool, appendOpenWhiskPath bool, encodeBodyAs string, useAuthentication bool) (*http.Request, error) {
+    return &http.Request{}, nil
+}
+
+func (c *MockClient) NewRequest(method, urlStr string, body interface{}, includeNamespaceInUrl bool) (*http.Request, error) {
+    actionRequest.Method = method
+    actionRequest.URL = urlStr
+
+    request, err := http.NewRequest(method, urlStr, nil)
+    if (err != nil) {
+        return &http.Request{}, err
+    }
+
+    return request, nil
+}
+
+func (c *MockClient) Do(req *http.Request, v interface{}, ExitWithErrorOnTimeout bool, secretToObfuscate ...ObfuscateSet) (*http.Response, error) {
+    var reader = strings.NewReader(actionResponse.Body)
+
+    dc := json.NewDecoder(reader)
+    dc.UseNumber()
+    err := dc.Decode(v)
+
+    if err != nil {
+        return nil, err
+    }
+
+    resp := &http.Response{
+        StatusCode: 200,
+        Body: ioutil.NopCloser(reader),
+    }
+
+    return resp, nil
+}
+
+func TestActionGet(t *testing.T) {
+    assert := assert.New(t)
+    mockClient := &MockClient{}
+    actionService := &ActionService{client: mockClient}
+
+    actionResponse.Body = NODE_ACTION_NO_CODE
+    action, _, _ := actionService.Get("test", false)
+
+    var exec Exec
+    exec = *action.Exec
+    var nilStr *string
+
+    assert.Equal("GET", actionRequest.Method)
+    assert.Equal("actions/test?code=false", actionRequest.URL)
+    assert.Equal(nilStr, exec.Code)
+
+    actionResponse.Body = NODE_ACTION
+    action, _, _ = actionService.Get("test", true)
```

Comment:
additional tests needed?
- NODE_ACTION with `actionService.Get("test", false)`
- NODE_ACTION_NO_CODE with `actionService.Get("test", true)`

might need to consider augmenting this unit test with some integration tests against a live controller to ensure the api contract is maintained.

Reply:
`NODE_ACTION` would not work with `actionService.Get("test", false)`. Same for the other one.

Reply:
Would it not compile or fail in an expected way?

Reply:
`NODE_ACTION` containers a code property and `actionService.Get("test", false)` says to not fetch the code.